src/main/java/com/pusher/client/connection/websocket/WebSocketConnection.java

```diff
@@ -251,21 +256,21 @@ public void run() {
 
     @Override
     public void onClose(final int code, final String reason, final boolean remote) {
-        activityTimer.cancelTimeouts();
+        if (state != ConnectionState.DISCONNECTED) {
+            activityTimer.cancelTimeouts();
 
-        factory.queueOnEventThread(new Runnable() {
-            @Override
-            public void run() {
-                if (state != ConnectionState.DISCONNECTED) {
+            factory.queueOnEventThread(new Runnable() {
+                @Override
+                public void run() {
                     updateState(ConnectionState.DISCONNECTED);
+                    factory.shutdownThreads();
                 }
-                else {
-                    log.error("Received close from underlying socket when already disconnected. " + "Close code ["
-                            + code + "], Reason [" + reason + "], Remote [" + remote + "]");
-                }
-                factory.shutdownThreads();
-            }
-        });
+            });
+        } else {
+            // Sometimes expected when we disconnected due to pong timeout
+            log.error("Received close from underlying socket when already disconnected. " + "Close code ["
+                    + code + "], Reason [" + reason + "], Remote [" + remote + "]");
+        }
```

Comment:
I think a short circuit exit at the top would be much clearer than wrapping the entire effective body of the method in an `if`.

Also, we should not log at `error` level in expected circumstances.